### PR TITLE
Option to mount a new local storage disabled

### LINKED ISSUE #26653: Make it possible to disable local storage mount type
In some scenarios, the ownCloud admin is not the same person as the sysadmin.
In such scenarios, the ownCloud admin only has access to the web interface and cannot run CLI commands and is not supposed to do anything outside of the OC folder.
Allowing such admin to mount arbitrary folders with the "Local" external storage could be considered a security issue in such specific scenarios.

So the idea would be to introduce a config.php option (which is edited by the sysadmin) to prevent admins to mount external storages of the type "Local".

@hodyroff @jvillafanez @DeepDiver1975

### PR DESCRIPTION
**Pull request by noveens**
Fixes #26653 

## Description

Option to mount a new local storage removed unless ```files_external_allow_create_new_local``` variable specifically set to ```true``` in config.php

## Related Issue

https://github.com/owncloud/core/issues/26653

## How Has This Been Tested?

Included the ```files_external_allow_create_new_local``` option and set to ```true``` and ```false```, works as desired

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

@PVince81

### CODE REVIEW

**Comment by noveens** [comment]
To discuss : The colored status indicator at the ```http://localhost/core/index.php/settings/admin?sectionid=storage``` page always turns to RED even if being displayed at the home page.

@PVince81 ?

**Comment by PVince81** [comment]
Are these two checkbox items related to your change or do you observe these also on master ?
What log entries ?

**Comment by noveens** [comment]
@PVince81 ,
The two checkbox items are related to my change, 
And I needed to discuss with you how to deal with these changes.

The logs are : 
```
{"reqId":"E1qVs3bsCjnSHqDjNdaz","level":3,"time":"2017-04-06T23:16:10+00:00","remoteAddr":"127.0.0.1","user":"--","app":"PHP","method":"GET","url":"\/core\/cron.php","message":"Undefined index: size at \/var\/www\/core\/lib\/private\/Files\/Cache\/Scanner.php#419"}
{"reqId":"E1qVs3bsCjnSHqDjNdaz","level":3,"time":"2017-04-06T23:16:10+00:00","remoteAddr":"127.0.0.1","user":"--","app":"PHP","method":"GET","url":"\/core\/cron.php","message":"Undefined index: size at \/var\/www\/core\/lib\/private\/Files\/Cache\/Scanner.php#416"}
{"reqId":"E1qVs3bsCjnSHqDjNdaz","level":3,"time":"2017-04-06T23:16:10+00:00","remoteAddr":"127.0.0.1","user":"--","app":"PHP","method":"GET","url":"\/core\/cron.php","message":"Undefined index: size at \/var\/www\/core\/lib\/private\/Files\/Cache\/Scanner.php#419"}
{"reqId":"E1qVs3bsCjnSHqDjNdaz","level":3,"time":"2017-04-06T23:16:10+00:00","remoteAddr":"127.0.0.1","user":"--","app":"PHP","method":"GET","url":"\/core\/cron.php","message":"Undefined index: size at \/var\/www\/core\/lib\/private\/Files\/Cache\/Scanner.php#416"}
{"reqId":"E1qVs3bsCjnSHqDjNdaz","level":3,"time":"2017-04-06T23:16:10+00:00","remoteAddr":"127.0.0.1","user":"--","app":"PHP","method":"GET","url":"\/core\/cron.php","message":"Undefined index: size at \/var\/www\/core\/lib\/private\/Files\/Cache\/Scanner.php#419"}
{"reqId":"E1qVs3bsCjnSHqDjNdaz","level":3,"time":"2017-04-06T23:16:10+00:00","remoteAddr":"127.0.0.1","user":"--","app":"PHP","method":"GET","url":"\/core\/cron.php","message":"Undefined index: size at \/var\/www\/core\/lib\/private\/Files\/Cache\/Scanner.php#416"}
{"reqId":"E1qVs3bsCjnSHqDjNdaz","level":3,"time":"2017-04-06T23:16:10+00:00","remoteAddr":"127.0.0.1","user":"--","app":"PHP","method":"GET","url":"\/core\/cron.php","message":"Undefined index: size at \/var\/www\/core\/lib\/private\/Files\/Cache\/Scanner.php#419"}
{"reqId":"E1qVs3bsCjnSHqDjNdaz","level":3,"time":"2017-04-06T23:16:10+00:00","remoteAddr":"127.0.0.1","user":"--","app":"PHP","method":"GET","url":"\/core\/cron.php","message":"Undefined index: size at \/var\/www\/core\/lib\/private\/Files\/Cache\/Scanner.php#416"}
{"reqId":"E1qVs3bsCjnSHqDjNdaz","level":3,"time":"2017-04-06T23:16:10+00:00","remoteAddr":"127.0.0.1","user":"--","app":"PHP","method":"GET","url":"\/core\/cron.php","message":"Undefined index: size at \/var\/www\/core\/lib\/private\/Files\/Cache\/Scanner.php#419"}
{"reqId":"E1qVs3bsCjnSHqDjNdaz","level":3,"time":"2017-04-06T23:16:10+00:00","remoteAddr":"127.0.0.1","user":"--","app":"PHP","method":"GET","url":"\/core\/cron.php","message":"Undefined index: size at \/var\/www\/core\/lib\/private\/Files\/Cache\/Scanner.php#416"}
{"reqId":"E1qVs3bsCjnSHqDjNdaz","level":3,"time":"2017-04-06T23:16:10+00:00","remoteAddr":"127.0.0.1","user":"--","app":"PHP","method":"GET","url":"\/core\/cron.php","message":"Undefined index: size at \/var\/www\/core\/lib\/private\/Files\/Cache\/Scanner.php#419"}
{"reqId":"E1qVs3bsCjnSHqDjNdaz","level":3,"time":"2017-04-06T23:16:10+00:00","remoteAddr":"127.0.0.1","user":"--","app":"PHP","method":"GET","url":"\/core\/cron.php","message":"Undefined index: size at \/var\/www\/core\/lib\/private\/Files\/Cache\/Scanner.php#416"}
{"reqId":"E1qVs3bsCjnSHqDjNdaz","level":3,"time":"2017-04-06T23:16:10+00:00","remoteAddr":"127.0.0.1","user":"--","app":"PHP","method":"GET","url":"\/core\/cron.php","message":"Undefined index: size at \/var\/www\/core\/lib\/private\/Files\/Cache\/Scanner.php#419"}
{"reqId":"E1qVs3bsCjnSHqDjNdaz","level":3,"time":"2017-04-06T23:16:10+00:00","remoteAddr":"127.0.0.1","user":"--","app":"PHP","method":"GET","url":"\/core\/cron.php","message":"Undefined index: size at \/var\/www\/core\/lib\/private\/Files\/Cache\/Scanner.php#416"}
```

**Comment by PVince81** [comment]
Do you observe these log entries also when running on master ? I don't see how these relate to your changes. It could be another important bug you just discovered.

**Comment by noveens** [comment]
@PVince81 ,
These used to appear on the latest master after the changes of #26690
When you set the flag ```files_external_allow_local``` to false in config.php, these logs still show up.

And what about the red colored indicator?

File: apps/files_external/lib/Controller/StoragesController.php [hunk 1]
```diff
@@ -111,26 +111,30 @@ protected function createStorage(
...
-		} catch (\InvalidArgumentException $e) {
-			$this->logger->logException($e);
-			return new DataResponse(
-				[
-					'message' => (string)$this->l10n->t('Invalid backend or authentication mechanism class')
-				],
-				Http::STATUS_UNPROCESSABLE_ENTITY
-			);
-		}
+		$canCreateNewLocalStorage = \OC::$server->getConfig()->getSystemValue('files_external_allow_create_new_local', false);
```
Comment (PVince81) [inline apps/files_external/lib/Controller/StoragesController.php]: move this to `function create`.

here you are also preventing updates to happen, so existing local storages cannot be updated/saved (red indicator)

**Comment by PVince81** [comment]
- [ ] BUG: need to find a way to remove "Local" from the selection list altogether, maybe filter it out when populating the list

On my local instance I didn't see any log entries related to scanner, so it might be specific to your env.

**Comment by noveens** [comment]
@PVince81 ,
Can you have a look?

**Comment by noveens** [comment]
@PVince81 ,

Ready for a merge?

File: apps/files_external/lib/Controller/GlobalStoragesController.php [hunk 2]
```diff
@@ -87,33 +87,44 @@ public function create(
...
+			$response = $this->validate($newStorage);
+			if (!empty($response)) {
+				return $response;
+			}
 
-		$newStorage = $this->service->addStorage($newStorage);
+			$newStorage = $this->service->addStorage($newStorage);
 
-		$this->updateStorageStatus($newStorage);
+			$this->updateStorageStatus($newStorage);
```
Comment (PVince81) [inline apps/files_external/lib/Controller/GlobalStoragesController.php]: use tabs, not spaces

**Comment by PVince81** [comment]
Tested, works 👍 

@noveens please fix the tabs then we can merge this

File: apps/files_external/lib/Controller/GlobalStoragesController.php (hunk 2 above)
Comment (noveens) [inline apps/files_external/lib/Controller/GlobalStoragesController.php]: Sure, will do.

Comment (noveens) [inline apps/files_external/lib/Controller/GlobalStoragesController.php]: @PVince81 ,
These are tabs not spaces, I guess GitHub is just trying to show the changes indentation?

File: apps/files_external/lib/Controller/GlobalStoragesController.php [hunk 3]
```diff
@@ -87,33 +87,44 @@ public function create(
...
-			$applicableUsers,
-			$applicableGroups,
-			$priority
-		);
-		if ($newStorage instanceof DataResponse) {
-			return $newStorage;
-		}
+		$canCreateNewLocalStorage = \OC::$server->getConfig()->getSystemValue('files_external_allow_create_new_local', false);
+
+		if ($canCreateNewLocalStorage === true) {
```
Comment (butonic) [inline apps/files_external/lib/Controller/GlobalStoragesController.php]: Unfortunately, this also triggers for smb. As it seems it prevents adding any external storage. Also it returns null and no meaningful message is shown in the web ui. Will add an issue.

Comment (PVince81) [inline apps/files_external/lib/Controller/GlobalStoragesController.php]: see https://github.com/owncloud/core/issues/27621

**Comment by PVince81** [comment]
Regression found, fixed here: https://github.com/owncloud/core/issues/27621

**Pull request merged**